### DISCUSSION TITLE
Redefining default hotkeys

### DISCUSSION THREAD

**Discussion by Balzhur** [d2phap/ImageGlass#1702] [Q&A] [answered]
How do I change default hotkeys? They are not defined in the config, but hardcoded.
I've tried to do following
`  "MenuHotkeys": {
    "MnuDeleteFromHardDisk": [
      "D"
    ]
  },`

this works, but default action for 'D' is working too, so I ended up with two actions for a single key press...

**Comment by d2phap** [comment, marked as answer]
Your setting is correct, <kbd>D</kbd> is being used for `MnuOpenWith` so you will need to override it, too:
```json
// in igconfig.json
{
  "MenuHotkeys": {
    "MnuDeleteFromHardDisk": ["D"],
    "MnuOpenWith": []
  }
}
```

Here is the full list of the default hotkey for menu:

https://github.com/d2phap/ImageGlass/blob/816359325a66f57153ab2d3ebe48c74e5ebb5abe/Source/ImageGlass/FrmMain/FrmMain.Configs.cs#L42-L142

**Reply by JHuberman** [reply]
_could you please tell me where I can find these settings to change them.  I too would like to change the  arrow and page up & down keys to:

PgUp = zoom in
PgDn = zoom out
Up = scroll up
Dn = scroll down

Left is set correctly to "previous" and
Right is set corretly to "next"_

**Reply by d2phap** [reply]
Your menu names for zooming in/out are as below:

```json
// in igconfig.json
{
  "MenuHotkeys": {
    "MnuZoomIn": ["your new hotkey here"],
    "MnuZoomOut": ["your new hotkey here"],
  }
}
```

To get the correct hotkey string, you can open IG settings > Tools, click "Add" button to show the "Add new tool" dialog. In the "Hotkeys" section, click "Add hotkey" button. You can start pressing the hotkey you want, it will show the hotkey string, use it to fill in the `MenuHotkeys` in `igconfig.json` file.

**Comment by Balzhur** [comment]
Thanks, I actually found it by myself, went here to answer my own question, but you were faster :)
This works too:
`  "MenuHotkeys": {
    "MnuDeleteFromHardDisk": [
      "D"
    ],
    "MnuOpenWith": [
      "None"
    ]
  },
`

**Comment by FurkanGozukara** [comment]
ok i think i need to edit

```
        { nameof(MnuCopyImageData),         [new(Keys.Control | Keys.C)] },
        { nameof(MnuCopyFile),              [new(Keys.Control | Keys.Shift | Keys.C)] },
```



```
  "MenuHotkeys": {},
  "MouseClickActions": {
    "LeftDoubleClick": {
      "ToggleOn": {
        "Executable": "IG_AutoSetActualSize",
        "Arguments": []
      }
    },
```

so can you give me what to do ? @d2phap 

I want to copy image with ctrl+c and then be able to paste into folder with ctrl+v

![image](https://github.com/user-attachments/assets/32c2911b-93df-475b-92e4-70b888371558)

**Reply by d2phap** [reply]
```json
// in igconfig.json
{
  "MenuHotkeys": {
    "MnuCopyImageData": ["Ctrl+Shift+C"],
    "MnuCopyFile": ["Ctrl+C"],
  }
}
```

**Reply by FurkanGozukara** [reply]
> ```json
>   "MenuHotkeys": {
>     "MnuCopyImageData": ["Ctrl+Shift+C"],
>     "MnuCopyFile": ["Ctrl+C"],
>   }
> ```

working awesome thank you so much

**Reply by Valour549** [reply]
@d2phap Hi there. What is the syntax for binding two different hotkeys for the same action?

For example I want to bind both tilde and F11 to fullscreen.

```
"MnuFullScreen": ["`"],
"MnuFullScreen": ["F11"]
```
```"MnuFullScreen": ["`" | "F11"]```

neither of these worked.

**Reply by d2phap** [reply]
@Valour549 You should use a comma `,`
```
"MnuFullScreen": ["`" , "F11"]
```

**Reply by Valour549** [reply]
@d2phap Hi again.

How can I bind **Up** so that it pans up, and **Down** so that it pans down (while I'm zoomed in looking at an image)?

**Reply by d2phap** [reply]
@Valour549 
```json
"MenuHotkeys": {
  "MnuPanUp": ["Up"],
  "MnuPanDown": ["Down"]
},
```

**Comment by eugenesvk** [comment]
How would you bind `{ Const.IGTOOL_EXIFTOOL,            [new(Keys.X)] },`?
`    public const string IGTOOL_EXIFTOOL = "Tool_ExifGlass";`

Tried different name variants, but nothing seems to work
Found: it's in a separate `"ToolId": "Tool_ExifGlass",` field

**Reply by d2phap** [reply]
The tool is bound by `Executable` and `Argument`:
- `Executable` is the path of application (`.exe`) or protocol (`ms-photos:`)
- `Argument` is the parameter to pass to the "Executable"

For ExifGlass tool, the Executable is `exifglass` (you can pass the full path of ExifGlass.exe file), the Argument is `<file>` which is a macro of the path of current image.

In the "Command preview", it shows the full command line of the tool:
```
exifglass "C:\fake dir\photo.jpg"
```

To change the hotkey of a tool, you click on the "Add hotkey" button. All tools are parsed into JSON in igconfig.json file:

```json
"Tools": [
  {
    "ToolId": "Tool_ExifGlass",
    "ToolName": "ExifGlass - EXIF metadata viewer",
    "Executable": "exifglass",
    "Argument": "<file>",
    "IsIntegrated": true,
    "Hotkeys": [
      "X"
    ]
  }
],
```

<img width="2279" height="1563" alt="image" src="https://github.com/user-attachments/assets/b617c2a1-aa71-49f6-8531-2b45fd650fb5" />

**Reply by eugenesvk** [reply]
Thanks for the details, I was just trying to rebind the menu, and this tool was in the menu, so was trying to use the same logic as for the other menu items


By the way, have you thought about the ability to change keybinds by ctrl+right clicking on the menu and picking them on the GUI instead of having to edit configs? So like you can do for tools? (Same could be done for toolbar buttons)

**Reply by d2phap** [reply]
@eugenesvk Yes, I will add UI for hotkey setting.

**Comment by francisco-lafe** [comment]
EDIT. Found it here: https://github.com/d2phap/ImageGlass/discussions/1702#discussioncomment-10879231

Hello, what is the correct approach for Combo Keys? 
In this thread the two cases are:
1. Individual hotkey ("E")
2. Multi individual hotkeys ("E","D")
3. ComboKeys: [    "MnuCopyImageData": ["Ctrl+Shift+C"], )


The one that's missing is for "combo" keys, say CTRL+C etc.

How are those set? With the "pipe" character as the source code? `["SHIFT|C"]` ?

**Reply by d2phap** [reply]
Yea you can use `+` for combining keys: `Ctrl+C`, `Shift+C`,...
